### ISSUE TITLE
Mouse Wheel Events

### ISSUE THREAD

**Issue by marceltaeumel** [OpenSmalltalk/opensmalltalk-vm#41]
As documented on [squeak-dev](http://forum.world.st/Mouse-Wheel-Events-Step-2-of-3-EventSensor-tp4900411.html), Bert, Tim and I came up with a simple format for mouse-wheel events -- consistent with other (mouse) event types.

The low-level event array format is straightforward:

``` Smalltalk
{
   7. "EventTypeMouseWheel"
   stamp.
   deltaX. "VM converts OS values into multiples of 120 units per notch"
   deltaY. "(see above)"
   buttons. "like mouse event"
   modifiers. "like mouse event"
   nil. "unused"
   window index.
}
```

The VM should normalize wheel deltas to _120 units per notch_ as suggested in the [Mozilla Dev Docs](https://developer.mozilla.org/en-US/docs/Web/Events/mousewheel#Chrome). Having this, one could easily implement smooth scrolling in the image.

The image-side code in Squeak 5.1rc1 for mouse-wheel events is here:

``` Smalltalk
{
   EventSensor >> #processMouseWheelEvent:.
   EventSensor >> #processEvent:.
   HandMorph >> #processEvents.
   HandMorph >> #generateMouseWheelEvent:.
   HandMorph >> #handleEvent:.
   MouseWheelState >> #initialize. "The whole class. Raises wheelUp|Down|Left|Right every 120 units."
   MouseWheelEvent >> #initialize. "The whole class."
   Morph >> #handleMouseWheel:.
   Morph >> #handlesMouseWheel:.
   Morph >> #mouseWheel:.
   EventHandler >> #on:send:to:.
   EventHandler >> #handlesMouseWheel:.
   EventHandler >> #mouseWheel:fromMorph:.
} do: [:method | ToolSet browse: method methodClass selector: method selector].
```

At the time of writing, mouse-wheel events are synthesized from CTRL+up/down keyboard events. This hack could then be removed from the VM and the respective keyboard events freed for real use cases. Since the image-side code has been updated to use wheel events in `ScrollPane` etc., most applications would remain compatible after this change. One could also easily add a preference to the image and _synthesize CTRL+up/down_ events for 100% backwards compatibility.

To check whether it works, just disable the mouse-wheel event synthesis, clear `ScrollPane >> #scrollByKeyboard:`, and try to scroll some lists by wheel:

``` Smalltalk
HandMorph synthesizeMouseWheelEvents: false.
ScrollPane compile: 'scrollByKeyboard: evt\ ^ false' withCRs classified: 'event handling'.
```

Note that _scroll-by-keyboard_ can also work after such a change because enabling that aformentioned preference has an effect, of course, also if the use actually pressed CTRL+up/down on the keyboard.

**Comment by nicolas-cellier-aka-nice** [comment]
Mouse wheels have been implemented in this commit for OSX and unix VM:

https://github.com/OpenSmalltalk/opensmalltalk-vm/commit/897ef1725e32c2eb3d24e3402b2e95b114b8b28b

I'm also proposing https://github.com/OpenSmalltalk/opensmalltalk-vm/pull/461 for windows VM

Those events will be generated with a VM parameter saved with image.
In Squeak 5.3, see: `Smalltalk sendMouseWheelEvents: true.`.

So we are near to close this issue...

However, the linux and OSX commit do not report button and modifiers state correctly (correctly = like every other mouse event).
We have to change that.

Also, some devices (like MacBook trackpad) generate a flow of wheel events with very small deltas (far less than 120, I observed deltas down to 1!).
This delivers far too many events to image side, making Morphic not responsive.
Thus we shall better accumulate a few of those events at VM side until a threshold is reached.
I have hardcoded a threshold of 20 on windows VM (`WHEEL_DELTA/6`), it kind of works for me... (maybe 10 would give a greater sensitivity).

I will create another branch with fixes for OSX/linux.
